### PR TITLE
[R&D Machine Overhaul, Part 1] - Various machines no longer require R&D machines to function.

### PR DESCRIPTION
**Pull request by Timberpoes**
## About The Pull Request

Removes various R&D machine requirements including circuit imprinters, protolathes, destructive analyzer and the mechfabs. All machines that previously required an R&D console to function are standalone, including omni protolathes and omni imprinters.

Replaces "production"-class R&D consoles with civilian modular computers on all maps. There are no longer "production"-class consoles.

Removes the concept of "core" R&D consoles. R&D consoles only research and handle tech disks and stuff. All existing core consoles have been switched to their basic archetype.

Because it's a pain in the ass to manage, material reclaimation has been removed from the destructive analyzer. The destructive analyzer is now ONLY for destroying items and unlocking research nodes such as illegal and alien tech.

## Why It's Good For The Game

We can now work on porting R&D interfaces to tgui bit by bit.

As long as machines REQUIRED an R&D console to function, the R&D console could not be ported to tgui without also porting every machine that relied on it to tgui too.

In other words, it allows for atomised changes to R&D-class machines on a machine-by-machine basis.

## Changelog
:cl:
del: R&D Consoles are no longer required to operate other machinery including mechfabs, omni-protolathes, omni-imprinters and the destructive analyzer. These consoles have been removed and replaced with basic modular computers.
del: The destructive analyzer has lost its material reclaimation feature. It now only destroys items and unlocks techweb nodes.
/:cl:

### CODE REVIEW

**Comment by JenqaDev** [comment]
Good PR, now I can build mechfabs outside of science without having the annoying R&D link error.
From what I understood, are you also making the RD consoles normal again, not requiring to stupidly replace the board with another(because of some PR that wanted to later implement more changes but never seen them)?

**Comment by Timberpoes** [comment]
> 
> 
> Good PR, now I can build mechfabs outside of science without having the annoying R&D link error.
> From what I understood, are you also making the RD consoles normal again, not requiring to stupidly replace the board with another(because of some PR that wanted to later implement more changes but never seen them)?

I am not. It's more a semantic change for coders only.

What were previously "production" R&D consoles (those found in the R&D lab, Robotics or Engineering) have been replaced with ordinary Modular Computers.

The "core" R&D consoles are now just R&D consoles. There's no such thing as a "core" or "production" variant. It's simply a plain old R&D console. As such, the R&D consoles in the RD office, on the bridge and in the old station ruin/ghost spawner are still the only machines capable of doing research.

My intention with adding modular computers where the old "production" R&D consoles were was one half laziness, and one half intending to some day make a tgui modular computer/tablet app to replace the old R&D console functionality somewhat.

**Comment by JenqaDev** [comment]
I see now. Still good PR

**Comment by Tlaltecuhtli** [comment]
while you are touching rd consoles can you make ejecting disk put the disk in your hand instead of on top of the console?
here is an example of the syntax: 



```
**/obj/machinery/computer/nanite_cloud_controller/proc/eject(mob/living/user)
	if(!disk)
		return
	if(!istype(user) || !Adjacent(user) ||!user.put_in_active_hand(disk))
		disk.forceMove(drop_location())
	disk = null**

```

File: code/modules/research/destructive_analyzer.dm [hunk 1]
```diff
@@ -164,3 +126,99 @@ Note: Must be placed within 3 tiles of the R&D Console
...
 	return TRUE
+
+/obj/machinery/rnd/destructive_analyzer/ui_interact(mob/user)
+	. = ..()
+	var/datum/browser/popup = new(user, "destructive_analyzer", name, 900, 600)
+	popup.set_content(ui_deconstruct())
+	popup.open()
+
+/obj/machinery/rnd/destructive_analyzer/proc/ui_deconstruct()		//Legacy code
+	var/list/l = list()
```
Comment (Rohesie) [inline code/modules/research/destructive_analyzer.dm:137]: Please, use something better than `l`
`dat` or `msg` work if you want something short

File: code/modules/research/destructive_analyzer.dm [hunk 2]
```diff
@@ -164,3 +126,99 @@ Note: Must be placed within 3 tiles of the R&D Console
...
+			anything = TRUE
+
+		if (!anything)
+			l += "Nothing!"
+
+		l += "</div>"
+
+	for(var/i in 1 to length(l))
+		if(!findtextEx(l[i], RDSCREEN_NOBREAK))
+			l[i] += "<br>"
```
Comment (Rohesie) [inline code/modules/research/destructive_analyzer.dm:203]: Couldn't you just add `<br>` to the messages that don't have this?
Does this need to be made at runtime?

File: code/modules/research/destructive_analyzer.dm [hunk 3]
```diff
@@ -164,3 +126,99 @@ Note: Must be placed within 3 tiles of the R&D Console
...
+	. = l.Join("")
+	return replacetextEx(., RDSCREEN_NOBREAK, "")
+
+/obj/machinery/rnd/destructive_analyzer/Topic(raw, ls)
+	if(..())
+		return
```
Comment (Rohesie) [inline code/modules/research/destructive_analyzer.dm:209]:
```suggestion
	. = ..()
	if(.)
		return
```

File: code/modules/research/rdconsole.dm [hunk 4]
```diff
@@ -141,7 +97,7 @@ Nothing else in the console has ID requirements.
 			to_chat(user, "<span class='warning'>Machine cannot accept disks in that format.</span>")
 			return
 		to_chat(user, "<span class='notice'>You insert [D] into \the [src]!</span>")
-	else if(!(linked_destroy && linked_destroy.busy) && !(linked_lathe && linked_lathe.busy) && !(linked_imprinter && linked_imprinter.busy))
+	else
 		. = ..()
```
Comment (Rohesie) [inline code/modules/research/rdconsole.dm:101]:
```suggestion
		return
	return ..()
```

File: code/modules/research/destructive_analyzer.dm (hunk 1 above)
Comment (Timberpoes) [reply code/modules/research/destructive_analyzer.dm:137]: I can change it, but I wanted to avoid trawling through this dumpster fire of legacy code making style updates, especially when it comes to the Destructive Analyzer which should be fairly trivial to create a tgui interface for which would render this all moot.

File: code/modules/research/destructive_analyzer.dm (hunk 2 above)
Comment (Timberpoes) [reply code/modules/research/destructive_analyzer.dm:203]: I have absolutely no clue, it's all copy-pasted code ported over from the R&D console to make this machine standalone.

**Comment by Timberpoes** [comment]
I was hoping to avoid re-writing this code as Parts 2 to X would be updating the machines to use tgui when uncoupled from the R&D Console. I didn't intend for this to be a code improvement PR, just an atomised PR to prepare these machines for what would ostensibly be a second re-write for tgui.

**Comment by kopoba** [comment]
My GOD you are scientist nightmare!
Now to do tech research as scientist you need not to just replace circuit in lab but build whole new rnd console from scratch!

**Comment by Dennok** [comment]
> My GOD you are scientist nightmare!
> Now to do tech research as scientist you need not to just replace circuit in lab but build whole new rnd console from scratch!

Just take RD console
